### PR TITLE
Nearby selection for SubListSwapMoveSelector

### PR DESCRIPTION
**Pull request by yurloc**
1. The left subList is selected.
2. Its first element becomes the origin value.
3. A destination value is obtained from the distance matrix (index).
4. The right subList is randomly selected from all possible subLists beginning at the destination value's index.
   - Minimum and maximum subList size configuration is honored.

**Original selection order not supported by subList move selectors!!!**

<details>
<summary>
How to replicate CI configuration locally?
</summary>

Build Chain tool does "simple" maven build(s), the builds are just Maven commands, but because the repositories relates and depends on each other and any change in API or class method could affect several of those repositories there is a need to use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.
 
[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is a build tool which can be used on command line locally or in Github Actions workflow(s), in case you need to change multiple repositories and send multiple dependent pull requests related with a change you can easily reproduce the same build by executing it on Github hosted environment or locally in your development environment. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

- for <b>pull request checks</b>  
  Please add comment: <b>Jenkins retest this</b>

- for a <b>specific pull request check</b>  
  please add comment: <b>Jenkins (re)run [optaplanner|optaplanner-quickstarts|optaweb-vehicle-routing] tests</b>
  
- for a <b>full downstream build</b> 
  - for <b>jenkins</b> job: 
    please add comment: <b>Jenkins run fdb</b>
  - for <b>github actions</b> job: 
    add the label `run_fdb`

- for a <b>compile downstream build</b>  
  please add comment: <b>Jenkins run cdb</b>

- for a <b>full production downstream build</b>  
  please add comment: <b>Jenkins execute product fdb</b>

- for an <b>upstream build</b>  
  please add comment: <b>Jenkins run upstream</b>

- for <b>quarkus branch checks</b>  
  Run checks against Quarkus current used branch  
  Please add comment: <b>Jenkins run quarkus-branch</b>

- for a <b>quarkus branch specific check</b>  
  Run checks against Quarkus current used branch  
  Please add comment: <b>Jenkins (re)run [optaplanner|optaplanner-quickstarts|optaweb-vehicle-routing] quarkus-branch</b>

- for <b>quarkus main checks</b>  
  Run checks against Quarkus main branch  
  Please add comment: <b>Jenkins run quarkus-main</b>

- for a <b>specific quarkus main check</b>  
  Run checks against Quarkus main branch  
  Please add comment: <b>Jenkins (re)run [optaplanner|optaplanner-quickstarts|optaweb-vehicle-routing] quarkus-branch</b>

- for <b>quarkus lts checks</b>  
  Run checks against Quarkus lts branch  
  Please add comment: <b>Jenkins run quarkus-lts</b>

- for a <b>specific quarkus lts check</b>  
  Run checks against Quarkus lts branch  
  Please add comment: <b>Jenkins (re)run [optaplanner|optaplanner-quickstarts|optaweb-vehicle-routing] quarkus-lts</b>

- for <b>native checks</b>  
  Run native checks  
  Please add comment: <b>Jenkins run native</b>

- for a <b>specific native check</b>  
  Run native checks 
  Please add comment: <b>Jenkins (re)run [optaplanner|optaplanner-quickstarts] native</b>

- for <b>native lts checks</b>  
  Run native checks against quarkus lts branch
  Please add comment: <b>Jenkins run native-lts</b>

- for a <b>specific native lts check</b>  
  Run native checks against quarkus lts branch
  Please add comment: <b>Jenkins (re)run [optaplanner|optaplanner-quickstarts] native-lts</b>

</details>

### CI Status

 You can check OptaPlanner repositories CI status from [Chain Status webpage](https://kiegroup.github.io/optaplanner/).

<details>
<summary>
How to backport a pull request to a different branch?
</summary>

In order to automatically create a **backporting pull request** please add one or more labels having the following format `backport-<branch-name>`, where `<branch-name>` is the name of the branch where the pull request must be backported to (e.g., `backport-7.67.x` to backport the original PR to the `7.67.x` branch).

> **NOTE**: **backporting** is an action aiming to move a change (usually a commit) from a branch (usually the main one) to another one, which is generally referring to a still maintained release branch. Keeping it simple: it is about to move a specific change or a set of them from one branch to another.

Once the original pull request is successfully merged, the automated action will create one backporting pull request per each label (with the previous format) that has been added.

If something goes wrong, the author will be notified and at this point a manual backporting is needed.

> **NOTE**: this automated backporting is triggered whenever a pull request on `main` branch is labeled or closed, but both conditions must be satisfied to get the new PR created.
</details>

### CODE REVIEW

File: core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/list/SubListSelectorFactory.java [hunk 1]
```diff
@@ -93,12 +108,80 @@ private SubListSelector<Solution_> applyMimicRecording(HeuristicConfigPolicy<Sol
...
+        boolean randomSelection = resolvedSelectionOrder.toRandomSelectionBoolean();
+
+        NearbyDistanceMeter<?, ?> nearbyDistanceMeter =
+                configPolicy.getClassInstanceCache().newInstance(nearbySelectionConfig,
+                        "nearbyDistanceMeterClass", nearbySelectionConfig.getNearbyDistanceMeterClass());
+        // TODO Check nearbyDistanceMeterClass.getGenericInterfaces() to confirm generic type S is an entityClass
+        NearbyRandom nearbyRandom = NearbyRandomFactory.create(nearbySelectionConfig).buildNearbyRandom(randomSelection);
+
+        if (nearbySelectionConfig.getOriginSubListSelectorConfig() == null) {
+            throw new IllegalStateException("TODO");
```
Comment (triceo) [inline core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/list/SubListSelectorFactory.java:143]: TODO

File: core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/list/SubListSelectorFactory.java [hunk 2]
```diff
@@ -93,12 +108,80 @@ private SubListSelector<Solution_> applyMimicRecording(HeuristicConfigPolicy<Sol
...
+        });
+
+        nearbySelectionConfig.validateNearby(minimumCacheType, resolvedSelectionOrder);
+
+        boolean randomSelection = resolvedSelectionOrder.toRandomSelectionBoolean();
+
+        NearbyDistanceMeter<?, ?> nearbyDistanceMeter =
+                configPolicy.getClassInstanceCache().newInstance(nearbySelectionConfig,
+                        "nearbyDistanceMeterClass", nearbySelectionConfig.getNearbyDistanceMeterClass());
+        // TODO Check nearbyDistanceMeterClass.getGenericInterfaces() to confirm generic type S is an entityClass
```
Comment (triceo) [inline core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/list/SubListSelectorFactory.java:139]: TODO

File: core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/heuristic/selector/value/ValueSelectorFactory.java [hunk 3]
```diff
@@ -489,6 +489,8 @@ private ValueSelector<Solution_> applyNearbySelection(HeuristicConfigPolicy<Solu
                     (EntityIndependentValueSelector<Solution_>) valueSelector,
                     (EntityIndependentValueSelector<Solution_>) originValueSelector,
                     nearbyDistanceMeter, nearbyRandom, randomSelection);
+        } else if (nearbySelectionConfig.getOriginSubListSelectorConfig() != null) {
+            throw new IllegalArgumentException("TODO");
```
Comment (triceo) [inline core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/heuristic/selector/value/ValueSelectorFactory.java:493]: TODO

**Review by triceo — APPROVED** [review]
LGTM.
I am pointing out all the TODOs; I usually wouldn't, but you said yesterday that your plan was to have them resolved before merging.

File: core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/heuristic/selector/entity/EntitySelectorFactory.java [hunk 4]
```diff
@@ -180,6 +180,9 @@ private EntitySelector<Solution_> applyNearbySelection(HeuristicConfigPolicy<Sol
             NearbySelectionConfig nearbySelectionConfig, SelectionCacheType minimumCacheType,
             SelectionOrder resolvedSelectionOrder, EntitySelector<Solution_> entitySelector) {
         boolean randomSelection = resolvedSelectionOrder.toRandomSelectionBoolean();
+        if (nearbySelectionConfig.getOriginEntitySelectorConfig() == null) {
+            throw new IllegalArgumentException("TODO");
```
[the remark this answers is not in the document]
Comment (yurloc) [reply core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/heuristic/selector/entity/EntitySelectorFactory.java:184]: Will be addressed in a separate PR. I will either A) provide a meaningful message or B) change the config class API to make this condition impossible.

File: core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/list/SubListSelectorFactory.java (hunk 2 above)
Comment (yurloc) [reply core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/list/SubListSelectorFactory.java:139]: Duplicated from existing code. Will be eliminated in a deduplication step or linked to a Jira.

File: core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/list/NearSubListNearbySubListSelector.java [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+
+            // Next steps:
+            // 1. Select a destination near to the origin based on a random nearby index.
+            // 2. Select a random subList with destination being its first element.
+
+            Object nearbyElementEntity = null;
+            Integer nearbyElementListIndex = -1;
+            int availableListSize = -1;
+
+            // TODO What if MIN is 500? We could burn thousands of cycles before we hit a availableListSize >= 500!
```
[the remark this answers is not in the document]
Comment (yurloc) [reply core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/list/NearSubListNearbySubListSelector.java:189]: Good question. I've done some investigation and I no longer see this as an issue. I could see decreasing calculation speed as I increased the minimumSubListSize but the drop wasn't as dramatic as I imagined and I couldn't even attribute it to the `while` loop on the next line. I imagine that the score calculation drop was caused by the increased frequency of large list operations.

[the remark this answers is not in the document]
Comment (yurloc) [reply core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/list/NearSubListNearbySubListSelector.java:189]: But there is a risk of an endless loop. The endless loop is prevented in `RandomSubListSelector` but I forgot to replicate the check here. I'm going to fix that and remove the TODO.

**Pull request merged**